crates/cubecl-core/src/runtime_tests/cmma.rs

```diff
@@ -939,6 +942,321 @@ pub fn test_cmma_manual<R: Runtime, AB: CubeElement + Numeric, CD: CubeElement +
+    let lhs_data: Vec<f32> = (0..m)
+        .flat_map(|i| (0..k).map(move |j| e2m1::from_bits(((i + j) % 15) as u8 + 1).to_f32()))
+        .collect();
+    //println!("lhs: {lhs_data:?}");
+    let lhs = e2m1x2::from_f32_slice(&lhs_data);
+    let lhs_scales_data: Vec<S> = (0..m)
+        .flat_map(|i| (0..scales_factor).map(move |j| S::from_bits((i * 2 + j + 120) as u8)))
+        .collect();
+
+    // RHS: matrix where each element = (row_index * 3) + column_index, col-major
+    let rhs_data: Vec<f32> = (0..n)
+        .flat_map(|j| (0..k).map(move |i| e2m1::from_bits(((i + j) % 15) as u8 + 1).to_f32()))
+        .collect();
+    let rhs = e2m1x2::from_f32_slice(&rhs_data);
+    let rhs_scales_data: Vec<S> = (0..n)
+        .flat_map(|j| (0..scales_factor).map(move |i| S::from_bits((i * 3 + j + 120) as u8)))
+        .collect();
+
+    let lhs = client.create(AB::as_bytes(&lhs));
+    let lhs_scales = client.create(S::as_bytes(&lhs_scales_data));
+    let rhs = client.create(AB::as_bytes(&rhs));
+    let rhs_scales = client.create(S::as_bytes(&rhs_scales_data));
+    let out = client.empty(core::mem::size_of::<f32>() * m * n);
+
+    unsafe {
+        kernel_scaled::launch::<AB, f32, S, R>(
+            &client,
+            CubeCount::Static(1, 1, 1),
+            cube_dimensions,
+            TensorArg::from_raw_parts::<AB>(&lhs, &[k / 2, 1], &[m, k / 2], ab_line_size as u8),
+            TensorArg::from_raw_parts::<AB>(&rhs, &[k / 2, 1], &[n, k / 2], ab_line_size as u8),
+            TensorArg::from_raw_parts::<f32>(&out, &[n, 1], &[m, n], 1),
+            TensorArg::from_raw_parts::<S>(
+                &lhs_scales,
+                &[scales_factor, 1],
+                &[m, scales_factor],
+                1,
+            ),
+            TensorArg::from_raw_parts::<S>(
+                &rhs_scales,
+                &[scales_factor, 1],
+                &[n, scales_factor],
+                1,
+            ),
+            TensorArg::from_raw_parts::<f32>(&out, &[n, 1], &[m, n], 1),
+            m as u32,
+            n as u32,
+            k as u32,
+            scales_factor as u32,
+        )
+    };
+
+    // Calculate expected results (row-major order)
+    let mut expected = Vec::with_capacity(m * n);
+    for i in 0..m {
+        // For each output row
+        for j in 0..n {
+            // For each output column
+            let mut sum = 0.0;
+            for l in 0..k {
+                let l_scales = l / (k / scales_factor);
+
+                // Dot product over k-dimension
+                let lhs_val = lhs_data[i * k + l]; // LHS[i, l]
+                let lhs_scale = lhs_scales_data[i * scales_factor + l_scales].to_f32();
+                let rhs_val = rhs_data[j * k + l];
+                let rhs_scale = rhs_scales_data[j * scales_factor + l_scales].to_f32();
+                sum += lhs_val * lhs_scale * rhs_val * rhs_scale;
+            }
+            expected.push(sum);
+        }
+    }
```

Comment:
Could we extract that in a function? It's probably the same used in many tests?

Reply:
Not sure they're similar enough to be worth pulling into a function. The test setups in that module are all fairly specific and the code required to make it generic would probably end up being more complicated. Just in the new tests, manual and scaled use different matrix layouts (row/row for manual, row/col for the scaled one because it makes fp4 easier to handle if they're already in the right format for CUDA and don't need repacking). The one above needs to handle strided inputs, etc.

Reply:
Also one uses integers because it minimizes precision mismatch, the other one uses f32 because the small floats can't properly represent larger integers (or small integers in the case of e8m0)